### PR TITLE
Add Multi-categorical stype

### PR DESCRIPTION
**Pull request by yiweny**
Add `multicategorical` stype

### CODE REVIEW

File: torch_frame/data/mapper.py [hunk 1]
```diff
@@ -90,6 +89,61 @@ def backward(self, tensor: Tensor) -> pd.Series:
         return ser
 
 
+class MultiCategoricalTensorMapper(TensorMapper):
+    r"""Maps any multi-categorical series into an index representation, with
+    :obj:`-1` denoting missing values(NaN) and no value denoting not belonging
```
Comment (vid-koci) [inline torch_frame/data/mapper.py:94]:
```suggestion
    :obj:`-1` denoting missing values (NaN) and no value denoting not belonging
```

File: torch_frame/datasets/fake.py [hunk 2]
```diff
@@ -68,6 +68,18 @@ def __init__(
                     arr[1::2] = np.nan
                 df_dict[col_name] = arr
                 col_to_stype[col_name] = stype.categorical
+        if stype.multicategorical in stypes:
+            # TODO: Currently having multiple mulit-categorical columns
```
Comment (vid-koci) [inline torch_frame/datasets/fake.py:72]:
```suggestion
            # TODO: Currently having multiple multi-categorical columns
```

File: test/data/test_mapper.py [hunk 3]
```diff
@@ -38,6 +39,23 @@ def test_categorical_tensor_mapper():
     pd.testing.assert_series_equal(out, pd.Series(['A', 'B', None, None, 'B']))
 
 
+def test_multicategorical_tensor_mapper():
```
Comment (vid-koci) [inline test/data/test_mapper.py:42]: Can you also add a test for `None` handling?

**Review by vid-koci — APPROVED** [review]

File: torch_frame/data/multi_nested_tensor.py [hunk 4]
```diff
@@ -187,6 +190,25 @@ def _apply(self, fn: Callable[[Tensor], Tensor]) -> 'MultiNestedTensor':
 
         return out
 
+    def dim(self) -> int:
+        return 2
```
Comment (weihua916) [inline torch_frame/data/multi_nested_tensor.py:194]: Why 2? shouldn't it be 3?

File: torch_frame/data/dataset.py [hunk 5]
```diff
@@ -128,14 +136,18 @@ def __call__(
...
             for col in col_names:
                 out = self._get_mapper(col).forward(df[col], device=device)
                 xs_dict[stype].append(out)
         feat_dict = {
-            stype: torch.stack(xs, dim=1)
+            stype: (torch.stack(xs, dim=1)
+                    if not stype.is_stored_in_multi_nested_tensor else
+                    MultiNestedTensor.stack(xs, dim=1))
             for stype, xs in xs_dict.items()
         }
```
Comment (weihua916) [inline torch_frame/data/dataset.py:152]: stype.use_multi_nested_tensor

File: torch_frame/data/dataset.py [hunk 6]
```diff
@@ -196,7 +210,13 @@ def __init__(
             raise ValueError(f"The column(s) '{missing_cols}' are specified "
                              f"but missing in the data frame")
 
+        if target_col is not None and self.col_to_stype[
+                target_col] == torch_frame.multicategorical:
```
Comment (weihua916) [inline torch_frame/data/dataset.py:214]: Add parentheses for nicer formatting

File: torch_frame/data/dataset.py [hunk 7]
```diff
@@ -316,7 +336,15 @@ def materialize(self, device: Optional[torch.device] = None,
 
         # 1. Fill column statistics:
         for col, stype in self.col_to_stype.items():
-            self._col_stats[col] = compute_col_stats(self.df[col], stype)
+            if stype == torch_frame.multicategorical:
```
Comment (weihua916) [inline torch_frame/data/dataset.py:339]: multi_categorical

File: torch_frame/data/multi_nested_tensor.py [hunk 8]
```diff
@@ -187,6 +190,25 @@ def _apply(self, fn: Callable[[Tensor], Tensor]) -> 'MultiNestedTensor':
...
+        if dim == 0:
+            return self.num_rows
+        elif dim == 1:
+            return self.num_cols
+        else:
+            raise ValueError(f"Dim {dim} does not exist.")
+
+    @staticmethod
+    def stack(xs: List['MultiNestedTensor'],
+              dim: int = 0) -> 'MultiNestedTensor':
```
Comment (weihua916) [inline torch_frame/data/multi_nested_tensor.py:206]: how does it depend on `dim`?

Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:206]: Can you add a TODO so that I can properly handle this once I added concatenation functionality?

File: torch_frame/stype.py [hunk 9]
```diff
@@ -23,12 +23,18 @@ class stype(Enum):
     numerical = 'numerical'
     categorical = 'categorical'
     text_embedded = 'text_embedded'
+    multicategorical = 'multicategorical'
```
Comment (weihua916) [inline torch_frame/stype.py:26]: multi_categorical

File: torch_frame/stype.py [hunk 10]
```diff
@@ -23,12 +23,18 @@ class stype(Enum):
...
 
     @property
     def is_text_stype(self) -> bool:
         return self in [stype.text_embedded]
 
+    @property
+    def is_stored_in_multi_nested_tensor(self) -> bool:
```
Comment (weihua916) [inline torch_frame/stype.py:33]: `use_multi_nested_tensor`

Also, add a short doc-string for this.

**Review by weihua916 — APPROVED** [review]
Thanks!

File: torch_frame/data/multi_nested_tensor.py [hunk 11]
```diff
@@ -187,6 +190,25 @@ def _apply(self, fn: Callable[[Tensor], Tensor]) -> 'MultiNestedTensor':
 
         return out
 
+    def dim(self) -> int:
```
[the remark this answers is not in the document]
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:193]: +1 should be 3

File: torch_frame/data/multi_nested_tensor.py [hunk 12]
```diff
@@ -187,6 +190,25 @@ def _apply(self, fn: Callable[[Tensor], Tensor]) -> 'MultiNestedTensor':
...
+        return 2
+
+    def size(self, dim) -> int:
+        if dim == 0:
+            return self.num_rows
+        elif dim == 1:
+            return self.num_cols
+        else:
+            raise ValueError(f"Dim {dim} does not exist.")
```
Comment (akihironitta) [inline torch_frame/data/multi_nested_tensor.py:202]: nit: to be consistent with `torch.Tensor.size()`
```suggestion
            raise IndexError(f"Dimension out of range (expected to be in range of [0, 1], but got {dim}")
```

File: torch_frame/data/multi_nested_tensor.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (yiweny) [reply torch_frame/data/multi_nested_tensor.py:193]: Got it. Yes, we need this. Ideally we need to implement any function that is supported by `torch.Tensor`

File: torch_frame/data/dataset.py (hunk 7 above)
Comment (yiweny) [reply torch_frame/data/dataset.py:339]: I think it's called `multicategorical` inside Kumo. We should probably align those two.

File: torch_frame/data/multi_nested_tensor.py (hunk 8 above)
Comment (yiweny) [reply torch_frame/data/multi_nested_tensor.py:206]: Sure!

File: test/data/test_mapper.py (hunk 3 above)
Comment (yiweny) [reply test/data/test_mapper.py:42]: Sure

File: test/data/test_dataset.py [hunk 13]
```diff
@@ -96,6 +100,16 @@ def test_converter():
     assert len(tf) == len(dataset)
 
 
+def test_multicategorical_materialization():
+    data = {'a': ['A,B', 'B,C,A', '', 'B', None]}
+    df = pd.DataFrame(data)
+    dataset = Dataset(df, {'a': stype.multicategorical})
+    dataset.materialize()
```
Comment (weihua916) [inline test/data/test_dataset.py:107]: Can we somehow test the correctness of `dataset.tensor_frame`?

File: torch_frame/data/mapper.py [hunk 14]
```diff
@@ -90,6 +89,62 @@ def backward(self, tensor: Tensor) -> pd.Series:
...
 
 
+class MultiCategoricalTensorMapper(TensorMapper):
+    r"""Maps any multi-categorical series into an index representation, with
+    :obj:`-1` denoting missing values (NaN) and no value denoting not belonging
+    to any categories.
+    """
+    def __init__(
+        self,
+        categories: Iterable[Any],
```
Comment (weihua916) [inline torch_frame/data/mapper.py:99]: Can we add to doc-string what `categories` is?

File: test/data/test_dataset.py (hunk 13 above)
Comment (weihua916) [reply test/data/test_dataset.py:107]: You can use `multi_nested_tensor[i,j]` to get a tensor.

File: torch_frame/data/dataset.py (hunk 7 above)
Comment (weihua916) [reply torch_frame/data/dataset.py:338]: I see. I still think `multi_categorical` is better. It's not difficult to align internally.

**Review by weihua916 — APPROVED** [review]
Thanks! Left some comments. LGTM after addressing comments.

File: test/data/test_dataset.py [hunk 15]
```diff
@@ -85,7 +86,10 @@ def test_dataset_inductive_transform():
 
 
 def test_converter():
-    dataset = FakeDataset(num_rows=10).materialize()
+    dataset = FakeDataset(
+        num_rows=10,
+        stypes=[stype.categorical, stype.numerical,
+                stype.multi_categorical]).materialize()
```
Comment (rusty1s) [inline test/data/test_dataset.py:92]: can we name this `multicategorical`?

File: torch_frame/data/dataset.py [hunk 16]
```diff
@@ -63,6 +65,7 @@ class DataFrameToTensorFrameConverter:
             column name into stats. Available as :obj:`dataset.col_stats`.
         target_col (str, optional): The column used as target.
             (default: :obj:`None`)
+        sep (str): Separator for multi-categorical columns. (default: :obj:`,`)
```
Comment (rusty1s) [inline torch_frame/data/dataset.py:68]: this is very weird. `sep` needs to be a stat, and it is especially column-specific.

File: torch_frame/data/dataset.py [hunk 17]
```diff
@@ -128,14 +136,17 @@ def __call__(
         r"""Convert a given :obj:`DataFrame` object into :class:`TensorFrame`
         object."""
 
-        xs_dict: Dict[torch_frame.stype, List[Tensor]] = defaultdict(list)
+        xs_dict: Dict[torch_frame.stype,
+                      List[Union[Tensor,
+                                 MultiNestedTensor]]] = defaultdict(list)
```
Comment (rusty1s) [inline torch_frame/data/dataset.py:141]: Can we group this as `TensorData = Union[Tensor, MultiNestedTensor]` (or similar)  in `typing.py`.

File: torch_frame/data/dataset.py [hunk 18]
```diff
@@ -316,7 +335,15 @@ def materialize(self, device: Optional[torch.device] = None,
 
         # 1. Fill column statistics:
         for col, stype in self.col_to_stype.items():
-            self._col_stats[col] = compute_col_stats(self.df[col], stype)
+            if stype == torch_frame.multi_categorical:
+                ser = self.df[col].apply(
+                    lambda x: [cat.strip() for cat in x.split(self.sep)]
+                    if x is not None and x != '' else [])
```
Comment (rusty1s) [inline torch_frame/data/dataset.py:341]: this needs to happen in tensor mapper, right?

File: torch_frame/data/mapper.py [hunk 19]
```diff
@@ -90,6 +89,68 @@ def backward(self, tensor: Tensor) -> pd.Series:
...
 
 
+class MultiCategoricalTensorMapper(TensorMapper):
+    r"""Maps any multi-categorical series into an index representation, with
+    :obj:`-1` denoting missing values (NaN) and no value denoting not belonging
+    to any categories.
+
+    Args:
+        categories (List[Any]): A list of possible categories in the
+        multi-categorical column.
```
Comment (rusty1s) [inline torch_frame/data/mapper.py:99]: sorted by occurencce.

File: torch_frame/data/mapper.py [hunk 20]
```diff
@@ -90,6 +89,68 @@ def backward(self, tensor: Tensor) -> pd.Series:
...
+    def _row_to_tensor(self, row: str):
+        if row == '':
+            return []
+        elif row is None:
+            return [-1]
+        else:
+            return [
+                self.categories.index(s) if s in self.categories else -1
+                for s in row.split(self.sep)
+            ]
```
Comment (rusty1s) [inline torch_frame/data/mapper.py:121]: can we use pandas join here? The for-loop is super slow.

File: torch_frame/data/dataset.py (hunk 18 above)
Comment (yiweny) [reply torch_frame/data/dataset.py:341]: Yeah, but when we calculate column stats, where we calculate the necessary arguments for Tensor Mapper, we need to split the columns by separator so we know what is the occurrence of each category.

File: torch_frame/data/dataset.py (hunk 17 above)
Comment (yiweny) [reply torch_frame/data/dataset.py:141]: Sure

File: test/data/test_dataset.py (hunk 15 above)
Comment (yiweny) [reply test/data/test_dataset.py:92]: 🤣 Just discussed with @weihua916 yesterday and he thinks `multi_categorical` is better.

File: torch_frame/data/dataset.py (hunk 16 above)
Comment (yiweny) [reply torch_frame/data/dataset.py:68]: Can you elaborate more? Thanks~

Comment (rusty1s) [reply torch_frame/data/dataset.py:68]: The seperator is not necessarily unique across columns.

File: torch_frame/data/dataset.py (hunk 18 above)
Comment (rusty1s) [reply torch_frame/data/dataset.py:341]: I get this. But I would have introduced a new stat type that does this splitting internally, and requires `sep` as an argument.

File: torch_frame/data/dataset.py (hunk 16 above)
Comment (yiweny) [reply torch_frame/data/dataset.py:68]: Got it.

File: torch_frame/data/multi_nested_tensor.py (hunk 8 above)
Comment (rusty1s) [reply torch_frame/data/multi_nested_tensor.py:220]: btw: This is not a stack but a cat :)

Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:220]: Noted. I was planning to modify it once `MultiNestedTensor` is in.
